### PR TITLE
Bring the new API changes to main branch

### PR DESCRIPTION
**Pull request by CarlinLiao**
Created a completely new sketch and half-implementation for aimsim's API.

Looking to have the code and general logic of the simulation flow reviewed before merging into master.

### CODE REVIEW

**Review by ribsthakkar — CHANGES_REQUESTED** [review]
In general this is a very well laid out foundation. Some of the changes I do recommend are using explicit imports rather than relative ones. I only commented in one place but my comment applies to nearly every file. Next, I think you need to make your variable names more clear and expliclty what they represent. No need to use `sd` and write a comment saying this is stopping distance when you can just write `stopping_dist`. In the two behemoth files of `lanes.py` and `intersections.py`, it was really hard to follow all of the logic where so much was going on with calculating reservations and all. This also leads to my next point, which is to really break those large functions down. Make a bunch of functions that represent each step. You have a bunch of comments explain what each section of code does. Those comments should be your function names/function calls and they should be implemented separately.  If you do that, it will read like a book, and if someone is interested in seeing how a certain part of the reservation calculation is done, they can click on the function to see the details. Other than that, it is really good progress.

**Review by pyrito — CHANGES_REQUESTED** [review]
Overall, I thought that this was really well laid out. The more modular the code is, the easier it will be to make modifications and test (and at this stage correctness is the most important thing). I think the next steps for you should be to start testing as you go. I recommend that you start writing out unit tests for the things that you have already implemented, otherwise it will become a major pain at the end and you won't get bogged down by design decisions.

**Review by ribsthakkar — COMMENTED** [review]
Aside from the one little typo I assume, the rest of the variable renaming and the imports are definitely better and more readable.
